### PR TITLE
Merges encoding branch

### PR DESCRIPTION
**Pull request by stevemk14ebr**
None of this is my work, just cherry picked alot from various forks and other  branches. Update to https://github.com/aquynh/capstone/pull/520 if this goes well i'll port to 'next' sometime. I tested these against code i had running under the encoding branch and all unit tests passed. I did not include all the commits of the encoding branch however and skipped a few:

such as:
https://github.com/aquynh/capstone/commit/f2a66cd5247ccf31271d3f4c62b84b124d60a8ee
https://github.com/aquynh/capstone/commit/062f9a8ac8587b53d015e0b576c6e8aa612c49af

also the modrm offset is not included as the branch i picked from didn't use it

i assume the python bindings are ok, i only tested the C. Assuming the fork i picked from is good then all should be well

### CODE REVIEW

File: arch/X86/X86Disassembler.c [hunk 1]
```diff
@@ -759,12 +759,23 @@ static void update_pub_insn(cs_insn *pub, InternalInstruction *inter, uint8_t *p
...
-	pub->detail->x86.disp = inter->displacement;
+	//pub->detail->x86.encoding.modrm_offset = inter->modRMOffset;
 
+	pub->detail->x86.sib = inter->sib;
 	pub->detail->x86.sib_index = x86_map_sib_index(inter->sibIndex);
 	pub->detail->x86.sib_scale = inter->sibScale;
 	pub->detail->x86.sib_base = x86_map_sib_base(inter->sibBase);
+
+	pub->detail->x86.disp = inter->displacement;
+	if(inter->consumedDisplacement){
```
Comment (aquynh) [inline arch/X86/X86Disassembler.c]: for consistent coding style, please put a space between `if` and `(`

Comment (aquynh) [reply arch/X86/X86Disassembler.c]: and also a space between `)` and `{`

File: arch/X86/X86Disassembler.c [hunk 2]
```diff
@@ -759,12 +759,23 @@ static void update_pub_insn(cs_insn *pub, InternalInstruction *inter, uint8_t *p
...
+		pub->detail->x86.encoding.disp_offset = inter->displacementOffset;
+		pub->detail->x86.encoding.disp_size = inter->displacementSize;
+	}
+
+	pub->detail->x86.encoding.imm_offset = inter->immediateOffset;
+	if (/*pub->detail->x86.encoding.imm_size == 0 && */inter->immediateOffset != 0){
```
Comment (aquynh) [inline arch/X86/X86Disassembler.c]: also a space here between `)` & `{`, please. (and elsewhere if i miss them)

File: arch/X86/X86DisassemblerDecoder.c [hunk 3]
```diff
@@ -2042,7 +2042,7 @@ static int readOperands(struct InternalInstruction *insn)
 					return -1;
 				// Apply the AVX512 compressed displacement scaling factor.
 				if (x86OperandSets[insn->spec->operands][index].encoding != ENCODING_REG && insn->eaDisplacement == EA_DISP_8)
-					insn->displacement *= 1 << (x86OperandSets[insn->spec->operands][index].encoding - ENCODING_RM);
+					insn->displacement *= (int64_t)1 << (x86OperandSets[insn->spec->operands][index].encoding - ENCODING_RM);
```
Comment (aquynh) [inline arch/X86/X86DisassemblerDecoder.c]: looks like this is a fix (for warning?) unrelated to this new feature?

Comment (stevemk14ebr) [reply arch/X86/X86DisassemblerDecoder.c]: from https://github.com/aquynh/capstone/commit/2d5b0a3308489896d5b08595e67db92a675a4f71 imho it should be 1ULL, ill fix it tomorrow with the other formatting

Comment (aquynh) [reply arch/X86/X86DisassemblerDecoder.c]: looks like this is because `displacement` type is changed from int32 to int64. why is this necessary?

Comment (stevemk14ebr) [reply arch/X86/X86DisassemblerDecoder.c]: i assumed it was as it was in the encoding branch, it very well may not be

File: include/x86.h [hunk 4]
```diff
@@ -203,6 +203,19 @@ typedef struct cs_x86_op {
 		bool avx_zero_opmask;
 } cs_x86_op;
 
+typedef struct cs_x86_encoding {
+	// ModR/M offset, or 0 when irrelevant
+	//uint8_t modrm_offset;
```
Comment (aquynh) [inline include/x86.h]: will we remove these commented lines above?

Comment (stevemk14ebr) [reply include/x86.h]: yea probably best to do so for now, i dont use modrm and if someone wants it later then can add it again from encoding. Or maybe i'll do it later when i merge to next

**Comment by aquynh** [comment]
@oleavr, do you have any comments?

**Comment by oleavr** [comment]
@aquynh Need `modrm_offset` for Frida's needs, so this won't help us in its current state.

**Comment by stevemk14ebr** [comment]
ill see if i can add modrm logic for you

**Comment by aquynh** [comment]
> ill see if i can add modrm logic for you
>

Yes please do that.

**Comment by stevemk14ebr** [comment]
requested refactoring done. Did C and python tests. Added modrm logic. Please review

File: tests/test_x86.c [hunk 5]
```diff
@@ -246,6 +251,6 @@ static void test()
 int main()
 {
 	test();
-
+	getchar();
```
Comment (aquynh) [inline tests/test_x86.c]: would you remove this line?

**Comment by aquynh** [comment]
could you remove two bins under msvc/ ?

File: bindings/python/test_x86.py [hunk 6]
```diff
@@ -47,10 +47,19 @@ def print_string_hex(comment, str):
 
     # print modRM byte
     print("\tmodrm: 0x%x" % (insn.modrm))
-
+	
+	# print modRM offset
```
Comment (aquynh) [inline bindings/python/test_x86.py]: please use tabs for indentation in C code (not space)

Comment (aquynh) [reply bindings/python/test_x86.py]: sorry, i mean use spaces in Python code to indent.

File: tests/test_x86.c [hunk 7]
```diff
@@ -50,7 +50,10 @@ static void print_insn_detail(csh ud, cs_mode mode, cs_insn *ins)
 
 	printf("\taddr_size: %u\n", x86->addr_size);
 	printf("\tmodrm: 0x%x\n", x86->modrm);
+	printf("\tmodrm_offset: 0x%x\n", x86->encoding.modrm_offset);
```
Comment (aquynh) [inline tests/test_x86.c]: how about only printing these new fields when they are relevant?

File: arch/X86/X86DisassemblerDecoder.c [hunk 8]
```diff
@@ -2086,6 +2088,14 @@ static int readOperands(struct InternalInstruction *insn)
 			case ENCODING_Ia:
 				if (readImmediate(insn, insn->addressSize))
 					return -1;
+				/* Direct memory-offset (moffset) immediate will get mapped
+								   to memory operand later. We want the encoding info to
```
Comment (aquynh) [inline arch/X86/X86DisassemblerDecoder.c]: would you fix indentation of these comment lines?

File: bindings/python/capstone/x86.py [hunk 9]
```diff
@@ -56,7 +64,7 @@ class CsX86(ctypes.Structure):
         ('addr_size', ctypes.c_uint8),
         ('modrm', ctypes.c_uint8),
         ('sib', ctypes.c_uint8),
-        ('disp', ctypes.c_int32),
+        ('disp', ctypes.c_int64),
```
Comment (aquynh) [inline bindings/python/capstone/x86.py]: is this a mistake? there is no change on "disp" in x86.h, but it gets changed here.

File: tests/test_x86.c (hunk 5 above)
Comment (aquynh) [reply tests/test_x86.c]: this was removed, but looks like it comes back again now?

File: tests/test_x86.c (hunk 7 above)
Comment (aquynh) [reply tests/test_x86.c]: looks like changes got reverted here too (on only printing relevant fields)?

**Comment by aquynh** [comment]
also, the bins under msvc/ come back again now?

**Comment by stevemk14ebr** [comment]
Sorry visual studio is really annoying with things like that. It breaks my git flow with garbage all the time

File: bindings/python/test_x86.py (hunk 6 above)
Comment (aquynh) [inline bindings/python/test_x86.py]: please fix indentation here, with spaces in Python code (this is regression i think)

Comment (aquynh) [reply bindings/python/test_x86.py]: and print these fields only when they are relevant.

**Comment by stevemk14ebr** [comment]
sorry, had repo sync issues causing regressions on previous review items. It's all fixed now

File: bindings/python/test_x86.py [hunk 10]
```diff
@@ -47,10 +47,22 @@ def print_string_hex(comment, str):
 
     # print modRM byte
     print("\tmodrm: 0x%x" % (insn.modrm))
-
+    
+    # print modRM offset
+	if insn.modrm_offset != 0:
```
Comment (aquynh) [inline bindings/python/test_x86.py]: please fix indentation here.

File: tests/test_x86.c [hunk 11]
```diff
@@ -50,7 +50,19 @@ static void print_insn_detail(csh ud, cs_mode mode, cs_insn *ins)
 
 	printf("\taddr_size: %u\n", x86->addr_size);
 	printf("\tmodrm: 0x%x\n", x86->modrm);
+	if (x86->encoding.modrm_offset != 0) {
+	    printf("\tmodrm_offset: 0x%x\n", x86->encoding.modrm_offset);
```
Comment (aquynh) [inline tests/test_x86.c]: we are using tabs for indentation in C code, but looks like this is using spaces?

**Comment by aquynh** [comment]
since we are at this, would you please also pick the commits you mentioned (https://github.com/aquynh/capstone/commit/062f9a8ac8587b53d015e0b576c6e8aa612c49af & https://github.com/aquynh/capstone/commit/f2a66cd5247ccf31271d3f4c62b84b124d60a8ee)?

besides these, are we missing anything else?

**Comment by stevemk14ebr** [comment]
Ill fix the formatting, but im done otherwise. https://github.com/aquynh/capstone/commit/f2a66cd5247ccf31271d3f4c62b84b124d60a8ee was reverted later, which is why i didn't pick it first.

**Comment by aquynh** [comment]
merged, thanks!

@oleavr, can you confirm that this works for Frida now?

**Comment by stevemk14ebr** [comment]
I can at least confirm it passes all the unit tests of polyhook ( over 300 )

**Comment by aquynh** [comment]
oops, this is a mistake: i did not mean to merge this to master, but only to next branch.

@stevemk14ebr, can you make a pull req for the next branch?

**Pull request merged**